Improve encode performance

tmehnert opened the pull request:
Hi. I have looked at issue #1 and find out that the current htmlEncode can be improved in performance. First i have tried to include html-entities, which will work but i was not brave enogh to include it, because it bundles a lot of stuff into the views bundle. Also htmlEncode is a public API of the library and i don't wanted to change it's produced encoding, because it might break user tests. So i ended up by creating a "own" implementation again, to speed things up.

### Micro-Benchmarks
The speed improvement highly depends on the input and the underlying javascript engine. I have prepared some micro-benchmarks for you, that shows the new version should be faster. Just click to run it, these benchmarks will only run the htmlEncode function with some input.

- [encode attribute, no escape](https://jsben.ch/KTc48) 
- [encode attribute, with escapes](https://jsben.ch/FY2dX)
- [encode content, no escape](https://jsben.ch/OegRF)
- [encode content, with escapes](https://jsben.ch/WTirT)

### Complate-cpp Benchmarks
I have also run complate-cpp rendering benchmarks with the current and the optimized implementation. I'm using a Intel i7-12700KF which runs an Ubuntu 20.04 on WSL2. All results are mean values of all iterations, i have taken best of five runs each, rounded to one decimal.

| Test                         | Current        | Optimized    |
| -------------------- | ------------: | ------------: |
| QuickJS, no todos    |         92.7 us |        60.0 us |
| QuickJS, 10 todos    |          1.2 ms |      771.1 us |
| QuickJS, 100 todos  |        11.5 ms |         7.5 ms |
| V8, no todos            |        10.7 us |          9.1 us |
| V8, 10 todos            |      149.4 us |      123.0 us |
| V8, 100 todos          |       1.44 ms |       1.26 ms |

To run the complate-cpp benchmarks by yourself on Linux, assuming you have the dependencies installed on your system.
```shell
mkdir build
cd build
cmake -DBUILD_TESTS=on ..
make
test/lib/complate-tests QuickJsRendererBenchmark
test/lib/complate-tests V8RendererBenchmark
```

tmehnert committed:
improve encode performance

tmehnert committed:
remove duplicated branch for single quot in regex

FND commented on src/html.js:
```diff
@@ -80,14 +80,24 @@ export function HTMLString(str) {
+	const rx = attribute ? /[&<>'"]/g : /[&<>]/g;
+	let res;
+	// fast path for strings which don't need to be encoded
+	if(rx.test(str)) {
```
Is it really worth running `#test` in addition to `#replace`? It seems to me that might be slower overall than just using `#replace` all the time:

```javascript
let res = str.replace(rx, c => HTML_ENCODE_REPLACE[c]);
```

A [cursory check](https://jsben.ch/D8DXB) suggests that `#replace` is almost as fast as `#test` when there's no match.

tmehnert replied:
I think you're probably right, maybe I did a little too much magic. Let me run some benchmarks without fastpath.

tmehnert replied:
Thanks to your intention i have done more benchmarking and research, ... with unexpected results ;)
On Firefox our "without fastpath" approach seems to be the best (most simple & good speed). But when it comes to chrome, things get different.

### On Chrome
When the content needs to be encoded, the current complate implementation outperform every approach we discussed so far, specially for non-attributes where the second branch will not be executed. But it's the slowest when it comes to attributes which don't have to be encoded, which should be in most cases. 

Just click this [Benchmark](https://jsben.ch/1sfpZ) on Chrome. That would bench the best possible input & engine for the current implementation. To make it easier to try out different inputs, i have put the parameters on top of the setup block. 

### New approach (loop based)
I have done more research and tried a loop based approach from [html-entities](https://github.com/mdevils/html-entities) (MIT License) and adopted to our needs. This implemtation tends to be the fastest under various inputs & engines.

tmehnert replied:
I will provide a new commit, but not today. Time to sleep^^

FND commented on src/html.js:
```diff
@@ -183,7 +193,7 @@ function generateAttributes(params, { tag, log, _idRegistry }) {
 		default:
 			// cf. https://html.spec.whatwg.org/multipage/syntax.html#attributes-2
-			if(/ |"|'|>|'|\/|=/.test(name)) {
+			if(/[ "'>/=]/.test(name)) {
```
Good catch there; this should have been a character set from the start. 👍

tmehnert replied:
Thanks ;)

FND reviewed (commented):
Thanks for this! I appreciate your thoughtful approach and agree that a third-party library would be undesirable.

tmehnert committed:
optimize encoding, when str need to be encoded

Last version had some penalties, on the non-fastpath.
Also this implementation tends to be the stable in performance
across different javascript engines.

tmehnert commented:
I have done a new version based on the algorithm of  [html-entities](https://github.com/mdevils/html-entities). I would really like to present you a easy oneliner, but it seems to that the loop based version deliver more stable performance across various engines.

### Complate-Cpp Benchmarks
If ran the Benchmarks again with our new algorithms. Loop-Based is the one i have commited.

| Test                         | Without Fastpath | Loop-Based  |
| -------------------- | -----------------: | ------------: |
| QuickJS, no todos    |                66.4 us |        61.3 us |
| QuickJS, 10 todos    |              828.5 us |   773.8.1 us |
| QuickJS, 100 todos  |                 8.0 ms |         7.5 ms |
| V8, no todos            |                 9.6 us |          9.0 us |
| V8, 10 todos            |             138.1 us |      125.2 us |
| V8, 100 todos          |              1.37 ms |       1.21 ms |

I have also updated the [online benchmark](https://jsben.ch/xhVX4).

tmehnert commented:
@FND what is your opinion on that?

FND commented on src/html.js:
```diff
@@ -80,14 +80,35 @@ export function HTMLString(str) {
+			res += str.substring(lastIndex);
+		}
+	} else {
+		res = str;
```
Personally, I prefer the [bouncer pattern](http://rikschennink.nl/thoughts/the-bouncer-pattern/) to avoid excessive nesting, reduce action at a distance etc.

```javascript
if(!match) {
    return str;
}

let rx = attribute ? /[&<>'"]/g : /[&<>]/g;
…
return res;
```

tmehnert replied:
Hmm.. yeah it would better match the nature of the function. If it would be able to speak it would say:
> I have seen so many strings in my life, i expect there are no evill in it and then it's no more todo. Just take it.
> But when i found something, then i have to do a bit more to protect you.
😄

tmehnert replied:
> This looks pretty solid!

Btw. I have to thank. You have created an amazing project (i really love JSX) and spend your time to let me part of it ❤️

FND replied:
I'm glad you find it useful and are willing to contribute.

I should mention that I'm not currently using complate myself and thus might not always be able to provide a timely response. So just to be safe, you might wanna send a quick note (e.g. by creating an issue) before starting any larger efforts. Having said that, there are other people around here who are perfectly capable of providing feedback.

FND commented on src/html.js:
```diff
@@ -80,14 +80,35 @@ export function HTMLString(str) {
+	let res;
+	let match = rx.exec(str);
+	if(match) {
+		res = "";
```
It used to be the case that string concatenation was expensive, so people collected strings in an array which was eventually `#join`d. We _might_ eke out a little more performance that way?

tmehnert replied:
I have seen so many unexpected results, that it would be always worth a try.
According to [The Performance of JavaScript String Concat](https://medium.com/@zhongdongy/the-performance-of-javascript-string-concat-e52466ca2b3a) form 2017, i expect that we are not gaining performance as long as we are not throwing a whole html document in one single `#encodeHtml` call.

But it would be explainable that `#join` will outperform at some input, when we make three assumptions about an engine:
* Array is fast.
* Strings are stored in a immutable piece of memory.
* Substrings are stored as a non owning reference to the original string.

Then just put the strings together will result in creating a new string, copy old content and deal with the old string which have to be deleted (or marked for it) on each concatination. Where on the `#join` strategy the result string will be created only once at the end, where the length of the underlying piece of memory is known.

tmehnert replied:
Hmm... when we take the medium post, then might the `#generateAttributes` will profit from using string concatination instead of `#join`, becuase it's a small amount of short strings. I make a try on that idea.

tmehnert replied:
> Hmm... when we take the medium post, then might the #generateAttributes will profit from using string concatination instead of #join, becuase it's a small amount of short strings. I make a try on that idea.

Bad idea from me. Raising the context will only cause more questions. Let's stick on `#encodeHtml`.

FND replied:
It looks like string concatenation is indeed faster these days: https://jsben.ch/VBdEz - I'm not entirely sure how representative that sample really is, but it seems to be a negligible micro-optimization either way.

FND reviewed (commented):
This looks pretty solid!

I don't currently have sufficient brains left to fully grok all the details, but hoping to return to that soon. Relying on - and validating - an existing implementation's experience certainly seems like a good choice - especially in the face of an ever-changing engine landscape. Thanks again for your efforts and for sticking with it.

I'm happy to run further performance tests (→ `Array#join`) myself and also tweak some minor stylistic preferences (→ bouncer pattern, avoiding `const` because I like to buck the trend, that sorta thing) as soon as I get to it.

FND committed:
tweaked stylistic details

naming, bouncer pattern, `let` consistency

FND commented:
As noted earlier, I've tweaked a few details in a5830ea2cfc7c0b3adc5e8458e47a0aaaf644105 (see commit message); if you're fine with that, I think this is ready to go.

tmehnert commented:
Your change looks good to me, the names are much more better. I ran all my tests and all have passed. The bechmark shows a little drop on QuickJs (0.4ms  @ 100 todos), not noticeable on other engines. But with what we achieved, this seems to me irrelevant.

I agree, let's take this nice profit. 👍

FND commented:
✅ Squashed as fb7fd6bbbf1b70dc554418eef75301e9d117e75b and released as v0.16.10.

I'm surprised there's a difference to the previous version, but that's just me idly wondering whether that's just chance - but yeah, negligible either way.

Thank you.
